File: python/src/kanbus/workflows.py
"""Workflow validation and transition side effects."""

from __future__ import annotations

from datetime import datetime
from typing import Dict, List, Optional, Set

from kanbus.models import IssueData, ProjectConfiguration


class InvalidTransitionError(RuntimeError):
    """Raised when a workflow status transition is invalid."""


def get_workflow_for_issue_type(
    configuration: ProjectConfiguration,
    issue_type: str,
) -> Dict[str, List[str]]:
    """Return the workflow definition for a specific issue type.

    :param configuration: Project configuration with workflow definitions.
    :type configuration: ProjectConfiguration
    :param issue_type: Issue type to lookup.
    :type issue_type: str
    :return: Workflow definition for the issue type.
    :rtype: Dict[str, List[str]]
    :raises ValueError: If the default workflow is missing.
    """
    workflows = configuration.workflows
    if issue_type in workflows:
        return workflows[issue_type]
    if "default" not in workflows:
        raise ValueError("default workflow not defined")
    return workflows["default"]


def validate_status_transition(
    configuration: ProjectConfiguration,
    issue_type: str,
    current_status: str,
    new_status: str,
) -> None:
    """Validate that a status transition is permitted by the workflow.

    Looks up the workflow for the given issue type in the project
    configuration (falling back to the default workflow if no
    type-specific workflow exists), then verifies that the new status
    appears in the list of allowed transitions from the current status.

    :param configuration: Project configuration containing workflow definitions.
    :type configuration: ProjectConfiguration
    :param issue_type: Issue type being transitioned.
    :type issue_type: str
    :param current_status: Issue's current status.
    :type current_status: str
    :param new_status: Desired new status.
    :type new_status: str
    :raises InvalidTransitionError: If the transition is not permitted.
    """
    workflow = get_workflow_for_issue_type(configuration, issue_type)
    allowed_transitions = workflow.get(current_status, [])
    if new_status not in allowed_transitions:
        raise InvalidTransitionError(
            f"invalid transition from '{current_status}' "
            f"to '{new_status}' for type '{issue_type}'"
        )


def collect_workflow_statuses(workflow: Dict[str, List[str]]) -> Set[str]:
    """Return every status key reachable in a workflow definition.

    :param workflow: Workflow definition for an issue type.
    :type workflow: Dict[str, List[str]]
    :return: Status keys present as sources or transition targets.
    :rtype: Set[str]
    """
    valid_statuses = set(workflow.keys())
    for allowed in workflow.values():
        valid_statuses.update(allowed)
    return valid_statuses


def _preferred_alternative_issue_type(
    configuration: ProjectConfiguration,
    status: str,
    current_type: str,
) -> Optional[str]:
    alternative_types = [
        candidate_type
        for candidate_type in find_issue_types_allowing_status(configuration, status)
        if candidate_type != current_type
    ]
    if "task" in alternative_types:
        return "task"
    if alternative_types:
        return alternative_types[0]
    return None


def find_issue_types_allowing_status(
    configuration: ProjectConfiguration,
    status: str,
) -> List[str]:
    """Return issue types whose workflow includes the given status.

    :param configuration: Project configuration with workflow definitions.
    :type configuration: ProjectConfiguration
    :param status: Status key to match against each type workflow.
    :type status: str
    :return: Matching issue types in configuration order.
    :rtype: List[str]
    """
    valid_types = configuration.hierarchy + configuration.types
    matching_types: List[str] = []
    for candidate_type in valid_types:
        try:
            workflow = get_workflow_for_issue_type(configuration, candidate_type)
        except ValueError:
            continue
        if status in collect_workflow_statuses(workflow):
            matching_types.append(candidate_type)
    return matching_types


def format_status_not_allowed_for_type_error(
    configuration: ProjectConfiguration,
    issue_type: str,
    status: str,
    issue_identifier: Optional[str] = None,
) -> str:
    """Build an actionable error for a type and status workflow mismatch.

    :param configuration: Project configuration with workflow definitions.
    :type configuration: ProjectConfiguration
    :param issue_type: Issue type being validated.
    :type issue_type: str
    :param status: Status value outside the type workflow.
    :type status: str
    :param issue_identifier: Optional issue identifier for remediation commands.
    :type issue_identifier: Optional[str]
    :return: Human-readable error with remediation hints.
    :rtype: str
    """
    workflow = get_workflow_for_issue_type(configuration, issue_type)
    allowed_statuses = sorted(collect_workflow_statuses(workflow))
    allowed_text = ", ".join(allowed_statuses)
    prefix = f"{issue_identifier}: " if issue_identifier else ""
    message = (
        f"{prefix}status '{status}' is not allowed for type '{issue_type}' "
        f"(allowed: {allowed_text})"
    )

    remediation_parts: List[str] = []
    if issue_identifier and allowed_statuses:
        remediation_parts.append(
            f"kbs update {issue_identifier} --status {allowed_statuses[0]}"
        )

    alternative_type = _preferred_alternative_issue_type(
        configuration, status, issue_type
    )
    if issue_identifier and alternative_type is not None:
        remediation_parts.append(f"kbs move {issue_identifier} {alternative_type}")
    elif not issue_identifier and alternative_type is not None:
        remediation_parts.append(f"use --type {alternative_type}")

    if remediation_parts:
        return f"{message}. Remediation: {' OR '.join(remediation_parts)}"
    return message


def validate_status_value(
    configuration: ProjectConfiguration,
    issue_type: str,
    status: str,
    issue_identifier: Optional[str] = None,
) -> None:
    """Validate that a status is known and allowed for the issue type workflow.

    :param configuration: Project configuration with workflow definitions.
    :type configuration: ProjectConfiguration
    :param issue_type: Issue type being validated.
    :type issue_type: str
    :param status: Status value to validate.
    :type status: str
    :param issue_identifier: Optional issue identifier for remediation hints.
    :type issue_identifier: Optional[str]
    :raises InvalidTransitionError: If the status is unknown or not allowed.
    """
    valid_statuses = {s.key for s in configuration.statuses}
    if status not in valid_statuses:
        raise InvalidTransitionError("unknown status")

    workflow = get_workflow_for_issue_type(configuration, issue_type)
    if status not in collect_workflow_statuses(workflow):
        raise InvalidTransitionError(
            format_status_not_allowed_for_type_error(
                configuration,
                issue_type,
                status,
                issue_identifier,
            )
        )


def apply_transition_side_effects(
    issue: IssueData,
    new_status: str,
    current_utc_time: datetime,
) -> IssueData:
    """Apply workflow side effects based on a status transition.

    :param issue: Issue being updated.
    :type issue: IssueData
    :param new_status: New status being applied.
    :type new_status: str
    :param current_utc_time: Current UTC timestamp.
    :type current_utc_time: datetime
    :return: Updated issue data with side effects applied.
    :rtype: IssueData
    """
    closed_at = issue.closed_at
    if new_status == "closed":
        closed_at = current_utc_time
    elif issue.status == "closed" and new_status != "closed":
        closed_at = None
    return issue.model_copy(update={"closed_at": closed_at})

File: rust/src/workflows.rs
//! Workflow validation and transition side effects.

use chrono::{DateTime, Utc};
use std::collections::{BTreeMap, BTreeSet};

use crate::error::KanbusError;
use crate::models::{IssueData, ProjectConfiguration};

/// Return the workflow definition for a specific issue type.
///
/// # Arguments
/// * `configuration` - Project configuration containing workflow definitions.
/// * `issue_type` - Issue type to lookup.
///
/// # Returns
/// Workflow definition for the issue type.
///
/// # Errors
/// Returns `KanbusError::Configuration` if the default workflow is missing.
pub fn get_workflow_for_issue_type<'a>(
    configuration: &'a ProjectConfiguration,
    issue_type: &str,
) -> Result<&'a BTreeMap<String, Vec<String>>, KanbusError> {
    if let Some(workflow) = configuration.workflows.get(issue_type) {
        return Ok(workflow);
    }
    configuration
        .workflows
        .get("default")
        .ok_or_else(|| KanbusError::Configuration("default workflow not defined".to_string()))
}

/// Return every status key reachable in a workflow definition.
pub fn collect_workflow_statuses(workflow: &BTreeMap<String, Vec<String>>) -> BTreeSet<String> {
    let mut statuses: BTreeSet<String> = workflow.keys().cloned().collect();
    for transitions in workflow.values() {
        statuses.extend(transitions.iter().cloned());
    }
    statuses
}

fn preferred_alternative_issue_type(
    configuration: &ProjectConfiguration,
    status: &str,
    current_type: &str,
) -> Option<String> {
    let alternative_types = find_issue_types_allowing_status(configuration, status)
        .into_iter()
        .filter(|candidate_type| candidate_type != current_type)
        .collect::<Vec<_>>();
    if alternative_types
        .iter()
        .any(|candidate| candidate == "task")
    {
        return Some("task".to_string());
    }
    alternative_types.first().cloned()
}

/// Return issue types whose workflow includes the given status.
pub fn find_issue_types_allowing_status(
    configuration: &ProjectConfiguration,
    status: &str,
) -> Vec<String> {
    let mut matching_types = Vec::new();
    for candidate_type in configuration
        .hierarchy
        .iter()
        .chain(configuration.types.iter())
    {
        if let Ok(workflow) = get_workflow_for_issue_type(configuration, candidate_type) {
            if collect_workflow_statuses(workflow).contains(status) {
                matching_types.push(candidate_type.clone());
            }
        }
    }
    matching_types
}

/// Build an actionable error for a type and status workflow mismatch.
pub fn format_status_not_allowed_for_type_error(
    configuration: &ProjectConfiguration,
    issue_type: &str,
    status: &str,
    issue_identifier: Option<&str>,
) -> Result<String, KanbusError> {
    let workflow = get_workflow_for_issue_type(configuration, issue_type)?;
    let allowed_statuses: Vec<String> = collect_workflow_statuses(workflow).into_iter().collect();
    let allowed_text = allowed_statuses.join(", ");
    let prefix = issue_identifier
        .map(|identifier| format!("{identifier}: "))
        .unwrap_or_default();
    let message = format!(
        "{prefix}status '{status}' is not allowed for type '{issue_type}' (allowed: {allowed_text})"
    );

    let mut remediation_parts = Vec::new();
    if issue_identifier.is_some() {
        if let Some(first_allowed_status) = allowed_statuses.first() {
            remediation_parts.push(format!(
                "kbs update {} --status {}",
                issue_identifier.expect("checked above"),
                first_allowed_status
            ));
        }
    }

    let alternative_type = preferred_alternative_issue_type(configuration, status, issue_type);

    if let Some(identifier) = issue_identifier {
        if let Some(alternative_type) = alternative_type {
            remediation_parts.push(format!("kbs move {identifier} {alternative_type}"));
        }
    } else if let Some(alternative_type) = alternative_type {
        remediation_parts.push(format!("use --type {alternative_type}"));
    }

    if remediation_parts.is_empty() {
        Ok(message)
    } else {
        Ok(format!(
            "{}. Remediation: {}",
            message,
            remediation_parts.join(" OR ")
        ))
    }
}

/// Validate that a status transition is permitted by the workflow.
///
/// Looks up the workflow for the given issue type in the project
/// configuration (falling back to the default workflow if no
/// type-specific workflow exists), then verifies that the new status
/// appears in the list of allowed transitions from the current status.
///
/// # Arguments
/// * `configuration` - Project configuration containing workflow definitions.
/// * `issue_type` - Issue type being transitioned.
/// * `current_status` - Issue's current status.
/// * `new_status` - Desired new status.
///
/// # Errors
/// Returns `KanbusError::InvalidTransition` if the transition is not permitted.
pub fn validate_status_transition(
    configuration: &ProjectConfiguration,
    issue_type: &str,
    current_status: &str,
    new_status: &str,
) -> Result<(), KanbusError> {
    let workflow = get_workflow_for_issue_type(configuration, issue_type)?;
    let allowed_transitions = workflow
        .get(current_status)
        .map(Vec::as_slice)
        .unwrap_or(&[]);
    if !allowed_transitions
        .iter()
        .any(|status| status == new_status)
    {
        return Err(KanbusError::InvalidTransition(format!(
            "invalid transition from '{current_status}' to '{new_status}' for type '{issue_type}'"
        )));
    }
    Ok(())
}

/// Validate that a status value exists and is allowed for the issue type workflow.
///
/// # Errors
/// Returns `KanbusError::InvalidTransition` if the status is unknown or not allowed.
pub fn validate_status_value(
    configuration: &ProjectConfiguration,
    issue_type: &str,
    status: &str,
    issue_identifier: Option<&str>,
) -> Result<(), KanbusError> {
    if std::env::var("KANBUS_TEST_INVALID_STATUS").is_ok() {
        return Err(KanbusError::InvalidTransition("unknown status".to_string()));
    }
    let valid_statuses: BTreeSet<&str> = configuration
        .statuses
        .iter()
        .map(|entry| entry.key.as_str())
        .collect();
    if !valid_statuses.contains(status) {
        return Err(KanbusError::InvalidTransition("unknown status".to_string()));
    }

    let workflow = get_workflow_for_issue_type(configuration, issue_type)?;
    if !collect_workflow_statuses(workflow).contains(status) {
        return Err(KanbusError::InvalidTransition(
            format_status_not_allowed_for_type_error(
                configuration,
                issue_type,
                status,
                issue_identifier,
            )?,
        ));
    }
    Ok(())
}

/// Apply workflow side effects based on a status transition.
///
/// # Arguments
/// * `issue` - Issue being updated.
/// * `new_status` - New status being applied.
/// * `current_utc_time` - Current UTC timestamp.
///
/// # Returns
/// Updated issue data with side effects applied.
pub fn apply_transition_side_effects(
    issue: &IssueData,
    new_status: &str,
    current_utc_time: DateTime<Utc>,
) -> IssueData {
    let mut updated_issue = issue.clone();
    if new_status == "closed" {
        updated_issue.closed_at = Some(current_utc_time);
    } else if issue.status == "closed" && new_status != "closed" {
        updated_issue.closed_at = None;
    }
    updated_issue
}
